sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeViewModel.kt

```diff
@@ -12,26 +13,51 @@ import javax.inject.Inject
 
 @HiltViewModel
 class HomeViewModel @Inject constructor(
-    private val useCase: UseCase,
+    private val getModelsUseCase: GetModelsUseCase,
+    private val getFirstTimeLaunchPreferencesUseCase: GetFirstTimeLaunchPreferencesUseCase,
+    private val updateFirstTimeLaunchPreferencesUseCase: UpdateFirstTimeLaunchPreferencesUseCase,
     dispatchers: DispatchersProvider
 ) : BaseViewModel(dispatchers) {
 
     private val _uiModels = MutableStateFlow<List<UiModel>>(emptyList())
-    val uiModels: StateFlow<List<UiModel>>
-        get() = _uiModels
+
+    private val _firstTimeLaunch = MutableStateFlow(false)
+
+    val homeViewState: StateFlow<HomeViewState> = combine(
+        _uiModels,
+        _firstTimeLaunch,
+        showLoading
+    ) { uiModels, firstTimeLaunch, showLoading ->
+        HomeViewState(
+            uiModels = uiModels,
+            firstTimeLaunch = firstTimeLaunch,
+            showLoading = showLoading
+        )
+    }.stateIn(
+        scope = viewModelScope,
+        started = SharingStarted.WhileSubscribed(5000),
+        initialValue = HomeViewState()
+    )
```

Comment:
@lydiasama, I think we should avoid combining these streams together:
- `_uiModels`
- `_firstTimeLaunch`
- `showLoading`

If I'm not wrong, it's because whenever one of the streams was triggered (**emit**), the rest streams were triggered, too(**trigger** **latest** **value**).  And I'm pretty sure this `_firstTimeLaunch` stream, we required it only once time to show the toast. So, the streams should be independent of each other in this case. What do you think? 🙏 cc @nimblehq/android-chapter

Reply:
@Wadeewee I rethink about this and found out we passed the `uiModels` and `showLoading` variable to the `HomeScreenContent` like:

```kotlin
HomeScreenContent(
        uiModels = uiModels,
        showLoading = showLoading,
        onItemClick = viewModel::navigateToSecond
    )
```

So when whether the `uiModels` or `showLoading` have an update, it will recompose this `HomeScreenContent` again. I'm not sure about if we pass the `homeViewState` to the `HomeScreenContent` instead, will it recompose when there is a variable updated? We might need some suggestion from @kaungkhantsoe 🙋🏻

Reply:
> Will it recompose when there is a variable updated?

@lydiasama , Regarding recompose, that's why we don't pass any state to `HomeScreenContent` to make it `Stateless` function:

```kotlin
HomeScreenContent(
        uiModels = uiModels,
        showLoading = showLoading,
        onItemClick = viewModel::navigateToSecond
    )
```
And because of `HomeScreenContent` is `Stateless` function then, we need to collect state in `Stateful` function, which is the `parent/root` function -> `HomeScreen` to avoid `recomposition`

```kotlin
fun HomeScreen( //------Stateful Fucntion--------//
    viewModel: HomeViewModel = hiltViewModel(),
    navigator: (destination: AppDestination) -> Unit
) {
....
    val showLoading: IsLoading by viewModel.showLoading.collectAsState()
    val uiModels: List<UiModel> by viewModel.uiModels.collectAsState()
    
     HomeScreenContent(//------Stateless Fucntion--------//
        uiModels = uiModels,
        showLoading = showLoading,
        onItemClick = viewModel::navigateToSecond
    )
}
```

> I'm not sure about if we pass the homeViewState to the HomeScreenContent instead:

And Regarding this, I think it's the same:

```kotlin
val homeViewState: HomeViewState by viewModel.homeViewState.collectAsState()

---------------VS---------------

val showLoading: IsLoading by viewModel.showLoading.collectAsState()
val uiModels: List<UiModel> by viewModel.uiModels.collectAsState()

/**
* Both ways we keep and collect it in `Stateful function`(HomeScreen), 
* and then we pass only `value` to `Stateless function`(HomeScreenContent`)
*/

```
That's from my understanding, Please let me know if you have any suggestion @kaungkhantsoe cc @Tuubz 🙇‍♀️

Reply:
@Wadeewee I think it's the same too. So I would prefer to wrap it inside the `homeViewState` to raise the best practice about handling state here. What do you think?

Reply:
From the [official](documentation), what `combine` does is
> Returns a [Flow](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/index.html) whose values are generated with [transform](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/combine.html) function by combining the most recently emitted values by each flow.

And just like @Wadeewee said, we should normally use `combine` when the streams are dependent on each other. But from Compose view, like @lydiasama said `HomeScreenContent` will try to recompose. It doesn't mean all the contents of `HomeScreenContent` will recompose. Assume the following as an example:
```
HomeScreenContent(val text1: String, val text2: String) {
   Text(text1)
   Text(text2)
}
```
Let's say originally `text1=Hello` and `text2=World`. If we try to pass `text2=Team` to `HomeScreenContent`, it will try to recompose every child. If you check the recomposition counts of `Text(text1)` and `Text(text2)`, you will see `Text(text1)` skipped the recomposition and `Text(text2)` recomposed once. 

In summary, if a parent compose view got different value parameters, it will try to recompose every child compose views. But only the child compose views with different values will be recomposed and the other will skip the recomposition.

I tried to remove some items in our item list using `HomeScreenState` and without using `HomeScreenState`. The recomposition count is the same as follows: 🙏 

|With HomeScreenState|Without HomeScreenState|
|--|--|
|https://user-images.githubusercontent.com/32578035/215387523-027bef91-2e9e-4f84-bdaf-dce875d0c188.mov|https://user-images.githubusercontent.com/32578035/215387571-2bfb4a26-83f3-4333-a392-186f5f0ef06f.mov|

Reply:
@kaungkhantsoe I think so. So I think there's no difference if we use `HomeScreenState` or not. And also, I think the code is cleaner when using it because we can wrap every stream inside one parameter. What do you think? cc @Wadeewee @nimblehq/android-chapter

Reply:
There is a solution if we don't want to use `combine`. We can separate functions for setting each states by `copy`.